include/openthread/platform/ble.h

```diff
@@ -0,0 +1,278 @@
+ */
+otError otPlatBleEnable(otInstance *aInstance);
+
+/**
+ * Disable the Bluetooth Low Energy radio.
+ *
+ * When disabled, the BLE stack will flush event queues and not generate new
+ * events. The BLE peripheral is turned off or put into a low power sleep
+ * state. Any dynamic memory used by the stack should be released,
+ * but static memory may remain reserved.
+ *
+ * @param[in] aInstance  The OpenThread instance structure.
+ *
+ * @retval OT_ERROR_NONE        Successfully transitioned to disabled.
+ * @retval OT_ERROR_FAILED      The BLE radio could not be disabled.
+ *
+ */
+otError otPlatBleDisable(otInstance *aInstance);
+
+/****************************************************************************
+ * @section Bluetooth Low Energy GAP.
+ ***************************************************************************/
+
+/**
+ * Starts BLE Advertising procedure.
+ *
+ * The BLE device shall use undirected advertising with no filter applied.
+ * A single BLE Advertising packet must be sent on all advertising
+ * channels (37, 38 and 39).
+ *
+ * @note This function shall be used only for BLE Peripheral role.
+ *
+ * @param[in] aInstance  The OpenThread instance structure.
+ * @param[in] aInterval  The interval between subsequent advertising packets
+ *                       in OT_BLE_ADV_INTERVAL_UNIT units.
+ *                       Shall be within OT_BLE_ADV_INTERVAL_MIN and
+ *                       OT_BLE_ADV_INTERVAL_MAX range.
+ *
+ * @retval OT_ERROR_NONE           Advertising procedure has been started.
+ * @retval OT_ERROR_INVALID_STATE  BLE Device is in invalid state.
+ * @retval OT_ERROR_INVALID_ARGS   Invalid interval value has been supplied.
+ *
+ */
+otError otPlatBleGapAdvStart(otInstance *aInstance, uint16_t aInterval);
+
+/**
+ * Stops BLE Advertising procedure.
+ *
+ * @note This function shall be used only for BLE Peripheral role.
+ *
+ * @param[in] aInstance  The OpenThread instance structure.
+ *
+ * @retval OT_ERROR_NONE           Advertising procedure has been stopped.
+ * @retval OT_ERROR_INVALID_STATE  BLE Device is in invalid state.
+ *
+ */
+otError otPlatBleGapAdvStop(otInstance *aInstance);
+
+/**
+ * The BLE driver calls this method to notify OpenThread that a BLE Device has
+ * been connected.
+ *
+ * @param[in]  aInstance     The OpenThread instance structure.
+ * @param[in]  aConnectionId The identifier of the open connection.
+ *
+ */
+extern void otPlatBleGapOnConnected(otInstance *aInstance, uint16_t aConnectionId);
+
+/**
+ * The BLE driver calls this method to notify OpenThread that the BLE Device
+ * has been disconnected.
+ *
+ * @param[in]  aInstance     The OpenThread instance structure.
+ * @param[in]  aConnectionId The identifier of the closed connection.
+ *
+ */
+extern void otPlatBleGapOnDisconnected(otInstance *aInstance, uint16_t aConnectionId);
+
+/**
+ * Disconnects BLE connection.
+ *
+ * The BLE device shall indicate the user has terminated the connection.
```

Comment:
This requirement is unclear to me: is it that the BLE Peripheral device (i.e. OpenThread device) needs to indicate to a user that the BLE Peripheral device itself has terminated the connection? But the disconnection procedure is maybe just started and not completed yet.

Does the `otPlatBleGapOnDisconnected` get called later on when the disconnection is done? Maybe that's a better moment to indicate something to the user.

The device may only have LEDs and not a display - why is there a requirement here to indicate something to the user?

Reply:
@canisLupus1313 It looks like this comment hasn't been processed yet. 

Maybe the word "indicate" refers to a BLE Indication here? That wasn't clear to me. Could be clarified as "shall send a BLE Indication message ... " or so , if that is the case. Blinking an LED is also an indication to the user.

Reply:
@EskoDijk To propose new text in the comment.

Reply:
What about the following - this captures the original intent of the line and is more clear, referring to the Bluetooth spec terminology. See Part F 1.3. It is still phrased as a requirement because every function in `ble.h` has some specific requirement towards the platform code, so I see that makes sense to keep.

```
The BLE device shall use the Remote User Terminated Connection (0x13) reason code when disconnecting from the peer BLE device.
```